### PR TITLE
Find best match for quote selectors in PDFs across all pages (v2)

### PR DESCRIPTION
**Pull request by robertknight**
~~**Depends on https://github.com/hypothesis/client/pull/3736**~~

This is a second attempt at resolving #3705. Compared to the first version in https://github.com/hypothesis/client/pull/3706 this approach aims to perform better in long documents and where the difference between the captured quote and current PDF text is only in whitespace content.

----

As described in #3705, stopping a quote search as soon as we find an
approximate match could result in quotes anchoring to suboptimal matches
if the position selector is no longer accurate.

We want to balance finding the best match against making the search
efficient in long documents. The approach taken in this commit is to
continue to search pages in priority order, but instead of returning the
first "good enough" match (below `matchQuote`'s error threshold), return
the first exact match ignoring whitespace differences, or the nearest
match in the whole document otherwise.

We ignore whitespace differences because this is the main change between
the current version of PDF.js that we include in Hypothesis products, and
newer versions that we are trying to migrate to (see #3687). It is also
likely to be the main source of text differences between PDF.js and
other viewers.

 - Change `anchorQuote` to ignore whitespace differences when searching
   for approximate matches
 - Change `anchorQuote` to keep searching for the best match if it does
   not find a whitespace-insensitive exact match in the current page
 - Add tests for whitespace changes

Fixes https://github.com/hypothesis/client/issues/3705

### CODE REVIEW

**Comment by robertknight** [comment]
I've done a decent amount of testing across recently annotated documents in the Public group. All looks to be working as expected.

Some URLs where I observed annotations that fail to anchor if using the newest PDF.js version and the published client, but which do anchor using the client on this branch:

https://www.ilga.gov/legislation/101/HB/PDF/10100HB0246lv.pdf
https://ocw.mit.edu/courses/mathematics/18-01-single-variable-calculus-fall-2006/lecture-notes/lec1.pdf
http://remikalir.com/wp-content/uploads/2017/09/Barron-et-al-2013.pdf
https://hcommons.org/?get_group_doc=1003678/1629909863-sedgwick-1997-paranoid-reading-and-reparative-reading.pdf

File: src/annotator/anchoring/pdf.js [hunk 1]
```diff
@@ -360,17 +398,59 @@ async function anchorQuote(quoteSelector, positionHint) {
...
+          score: match.score,
+        },
+      };
+
+      // If we found an exact match, stop early. This improves search performance
+      // in long documents compared to testing every page.
+      //
+      // A known limitation here is that the quote context (prefix and suffix)
+      // is not considered.
+      if (strippedText.slice(match.start, match.end) === strippedQuote) {
```
Comment (esanzgar) [inline src/annotator/anchoring/pdf.js:445]: Why not to fix this limitation?

It could be  something like this, I believe:

```
      // If we found an exact match, stop early. This improves search performance
      // in long documents compared to testing every page.
      const isExactQuote =
        strippedText.slice(match.start, match.end) === strippedQuote;
      const isExactPrefix =
        strippedPrefix === undefined
          ? true
          : strippedText.slice(
              match.start - strippedPrefix.length,
              match.start
            ) === strippedPrefix;
      const isExactSuffix =
        strippedSuffix === undefined
          ? true
          : strippedText.slice(
              match.end,
              match.end + strippedSuffix.length + 1
            ) === strippedPrefix;
      if (isExactQuote && isExactPrefix && isExactSuffix) {
```

File: src/annotator/anchoring/pdf.js [hunk 2]
```diff
@@ -360,17 +398,59 @@ async function anchorQuote(quoteSelector, positionHint) {
...
+        match: {
+          start: offsets[match.start],
+
+          // `match.end` is the offset one past the last character of the match
+          // in the stripped text. We need the offset one past the corresponding
+          // character in the original text.
+          //
+          // We assume here that matches returned by `matchQuote` must have
+          // be non-empty, so `match.end` > `match.start`.
+          end: offsets[match.end - 1] + 1,
```
Comment (esanzgar) [inline src/annotator/anchoring/pdf.js:434]: One additional benefit of this approach is that trims leading and trailing whitespace from the original anchor, if any.

**Review by esanzgar — APPROVED** [review]
All looks good. I have adde a few comments and suggestions.

I am going to test with a couple of PDFs and then approve both, this PR and #3736.

**Review by esanzgar — CHANGES_REQUESTED** [review]
My testing has raised some situation where the anchoring logic presented in this PR results in incorrect anchoring for very short quotes.

These errors are fixed by the suggestion in this comment:
https://github.com/hypothesis/client/pull/3743#discussion_r706001943

Would you be willing to consider the suggested change?

File: src/annotator/anchoring/pdf.js (hunk 1 above)
Comment (robertknight) [reply src/annotator/anchoring/pdf.js]: Yes, I think this is a decent trade-off between keeping the heuristics simple, finding the best match and optimizing the search. I've added a commit (baebd97a72f636854b26ffdb7de1d9e54b13d22e) which applies the basic idea from this proposal but handles missing `prefix` and `suffix` fields slightly:

- If the quote selector has either field, then require a match in at least one of them
- If the quote selector does not have either field, then don't require a match

The code as written in the proposal will ignore the `prefix` if the selector has a suffix and vice-versa.

Comment (esanzgar) [reply src/annotator/anchoring/pdf.js]: 👍🏼

File: src/annotator/anchoring/pdf.js [hunk 3]
```diff
@@ -312,9 +312,38 @@ async function anchorByPosition(pageIndex, start, end) {
...
+    const char = str[i];
+    if (char === ' ' || char === '\t' || char === '\n') {
+      continue;
+    }
+    stripped += char;
+    offsets.push(i);
+  }
+
+  return [stripped, offsets];
+}
```
[the remark this answers is not in the document]
Comment (robertknight) [reply src/annotator/anchoring/pdf.js:335]: `stripSpaces` may be called many times during quote anchoring with a large amount of text in total (eg. up to say a million chars in total for a large book like the "Hobbes' Leviathan" text). For that reason I wrote this in a more performance-minded way than I write most code. You can see some benchmarks at https://esbench.com/bench/613f1ca06c89f600a570114b. The results vary quite a bit between browsers, but for the length of string that we expect to see here, the combination of simple char tests + string concatentation is consistently faster.

**Review by esanzgar — APPROVED** [review]
Robert, very good work dealing with the anchoring issues raised by the new PDF.js! I confirm this PR works well for in all test cases I checked. I will approve also #3736.

File: src/annotator/anchoring/pdf.js [hunk 4]
```diff
@@ -312,9 +312,38 @@ async function anchorByPosition(pageIndex, start, end) {
...
+ * @param {string} str
+ * @return {[string, number[]]}
+ */
+function stripSpaces(str) {
+  const offsets = [];
+  let stripped = '';
+
+  for (let i = 0; i < str.length; i++) {
+    const char = str[i];
+    if (char === ' ' || char === '\t' || char === '\n') {
```
[the remark this answers is not in the document]
Comment (robertknight) [reply src/annotator/anchoring/pdf.js:331]: The testing for only a small number of the most important space chars, as opposed to `\s` or a Unicode character class check, is a choice for performance. I added a comment about this in the function description.

File: src/annotator/anchoring/pdf.js [hunk 5]
```diff
@@ -312,9 +312,38 @@ async function anchorByPosition(pageIndex, start, end) {
...
+    offsets.push(i);
+  }
+
+  return [stripped, offsets];
+}
+
 /**
  * Search for a quote in the given pages.
  *
+ * When comparing quote selectors to document text, ASCII whitespace characters
```
[the remark this answers is not in the document]
Comment (robertknight) [reply src/annotator/anchoring/pdf.js:344]: Given the current use case, which is to be invariant to spaces generated by the PDF renderer, it's not necessary. We can always revise this in future if we decide that we have a good reason to need to ignore these spaces as well.

File: src/annotator/anchoring/pdf.js [hunk 6]
```diff
@@ -360,17 +398,59 @@ async function anchorQuote(quoteSelector, positionHint) {
...
     }
 
-    const match = matchQuote(text, quoteSelector.exact, {
-      prefix: quoteSelector.prefix,
-      suffix: quoteSelector.suffix,
-      hint,
+    const [strippedText, offsets] = stripSpaces(text);
+
+    // Convert expected offset in original text into offset into stripped text.
+    let strippedHint;
```
[the remark this answers is not in the document]
Comment (robertknight) [reply src/annotator/anchoring/pdf.js:408]: TypeScript does infer the type of `strippedHint` correctly and will complain if you try to use it in a place that expects a different type. Confusingly VS Code reports the type as `any` if you hover the declaration of the variable, but the expected inferred type if you hover it in places where it is read.

File: src/annotator/anchoring/pdf.js [hunk 7]
```diff
@@ -344,7 +373,16 @@ async function anchorQuote(quoteSelector, positionHint) {
     });
   }
 
-  // Search pages until we find a match.
+  // Search pages for the best match, ignoring whitespace differences.
+  const [strippedPrefix] = quoteSelector.prefix
+    ? stripSpaces(quoteSelector.prefix)
+    : [];
```
[the remark this answers is not in the document]
Comment (robertknight) [reply src/annotator/anchoring/pdf.js]: An empty prefix or suffix is not terribly useful as it always matches, but it seems sensible to make this change given that the branch is mainly testing "is the prefix / suffix a string?"

File: src/annotator/anchoring/pdf.js [hunk 8]
```diff
@@ -360,17 +398,59 @@ async function anchorQuote(quoteSelector, positionHint) {
...
+    const [strippedText, offsets] = stripSpaces(text);
+
+    // Convert expected offset in original text into offset into stripped text.
+    let strippedHint;
+    if (hint !== undefined) {
+      strippedHint = 0;
+      while (strippedHint < offsets.length && offsets[strippedHint] < hint) {
+        ++strippedHint;
+      }
+    }
```
[the remark this answers is not in the document]
Comment (robertknight) [reply src/annotator/anchoring/pdf.js]: Ah, indeed. This is nicer. I'll do this.
